### PR TITLE
LEAF 4061 - Optimize form/query and getCustomData's read access check (2 of 3)

### PR DESCRIPTION
**Pull request by mgaoVA**
Note the PR merge base. This series of optimizations have been split into multiple PRs to simplify review.

This modifies getCustomData() to avoid duplicate checkReadAccess() queries. If checkReadAccess() was already run, then it's not necessary for getCustomData() to call checkReadAccess() again because it could only check a subset of the already checked records.

### Potential Impact
Read access to records

### Testing
Preconditions:
- A record that the current user doesn't have access to read
- The record doesn't show up in the Report Builder
- [x] After this patch, the current user still cannot open the record, and cannot see it in the Report Builder

### CODE REVIEW

**Review by maxrichardjr — APPROVED** [review]

**Review by jampaul3 — APPROVED** [review]

File: LEAF_Request_Portal/sources/Form.php [hunk 1]
```diff
@@ -2302,10 +2302,16 @@ public function getChildForms($recordID)
...
+     * WARNING: $alreadyCheckedReadAccess can only be set to true if $recordID_list has been
+     *          processed by checkReadAccess().
      *
+     * @param array $recordID_list
+     * @param array $indicatorID_list
+     * @param bool (optional) $alreadyCheckedReadAccess
      * @return array on success | false on malformed input
      */
-    public function getCustomData(array $recordID_list, string|null $indicatorID_list)
+    public function getCustomData(array $recordID_list, string|null $indicatorID_list, bool $alreadyCheckedReadAccess = false)
```
Comment (Pelentan) [inline LEAF_Request_Portal/sources/Form.php:2314]: Needs return type

File: LEAF_Request_Portal/sources/Form.php [hunk 2]
```diff
@@ -3628,7 +3636,7 @@ public function query($inQuery)
                 $indicatorIDs .= $indicatorID . ',';
             }
 
-            return $this->getCustomData($data, $indicatorIDs);
+            return $this->getCustomData($data, $indicatorIDs, $alreadyCheckedReadAccess);
```
Comment (Pelentan) [inline LEAF_Request_Portal/sources/Form.php:3639]: Functions should have only one return.  Can this be turned to a variable and then passed back with the final return?

**Review by Pelentan — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/sources/Form.php (hunk 1 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/Form.php]: Updated.

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

File: LEAF_Request_Portal/sources/Form.php (hunk 2 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/Form.php]: Updated.

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Review by aerinkayne — APPROVED** [review]

**Review by raphaelfretter — APPROVED** [review]

**Review by AmandaSumner — APPROVED** [review]

**Pull request merged**